libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.tcc

```diff
@@ -363,6 +363,9 @@ r1cs_ppzksnark_keypair<ppT> r1cs_ppzksnark_generator(const r1cs_ppzksnark_constr
 
     libff::enter_block("Compute the H-query", false);
     libff::G1_vector<ppT> H_query = batch_exp(libff::Fr<ppT>::size_in_bits(), g1_window, g1_table, Ht);
+#ifdef USE_MIXED_ADDITION
```

Comment:
When libsnark is compiled with `USE_MIXED_ADDITION`, some functions (at least `multi_exp_with_mixed_addition` and `multi_exp<..,..,multi_exp_method_djb>`) assume that mixed addition can be performed between arbitrary group elements and the input elements. This means the input elements have to be in special form, and converting them to that form is what's happening in the `#ifdef`d blocks.

In one case, the existing code was *already* assuming that the input was in special form, even though it wasn't specifically converted. So the ifdef'd parts were actually already needed, and their absence was a bug. Unfortunately, I don't quite remember where that was.

In the rest of the cases, the conversion had to be added to switch from `multi_exp_method_bos_coster` to `multi_exp_method_djb` (because input in special form is needed by the latter, but not by the former).

If they are omitted, multiexponentiation might give wrong answers, but only when compiling with `USE_MIXED_ADDITION`.

`multi_exp_method_djb` could be modified to not assume that the input is in special form, but that would make provers a little bit slower. Given that there were already instances where input was being converted to special form like this (e.g. [here](https://github.com/scipr-lab/libsnark/blob/candidate-master/libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.tcc#L366-L368)), I figured it would not be a problem to add more.